source/fontloader.ts

```diff
@@ -0,0 +1,235 @@
+            ['id', 'x', 'y', 'width', 'height', 'xoffset', 'yoffset', 'xadvance']);
+
+        const index: number = parseInt(pairs.get('id')!, 10);
+        assert(index > 0, `handleChar: glyph index is not greater than 0.`);
+
+        const glyph = new Glyph();
+
+        glyph.index = index;
+
+        const extentScale: GLfloat2 = [
+            1.0 / fontFace.glyphTextureExtent[0],
+            1.0 / fontFace.glyphTextureExtent[1],
+        ];
+        const extent: GLfloat2 = [
+            parseFloat(pairs.get('width')!),
+            parseFloat(pairs.get('height')!),
+        ];
+
+        glyph.subTextureOrigin = [
+            parseFloat(pairs.get('x')!) * extentScale[0],
+            1.0 - (parseFloat(pairs.get('y')!) + extent[1]) * extentScale[1],
+        ];
+
+        glyph.extent = extent;
+
+        glyph.subTextureExtent[0] = extent[0] * extentScale[0];
+        glyph.subTextureExtent[1] = extent[1] * extentScale[1];
+
+        glyph.setBearing(fontFace.base,
+            parseFloat(pairs.get('xoffset')!),
+            parseFloat(pairs.get('yoffset')!),
+        );
+
+        glyph.advance = parseFloat(pairs.get('xadvance')!);
+
+        fontFace.addGlyph(glyph);
+    }
+
+    protected handleKerning(stream: Array<string>, fontFace: FontFace): void {
+        const pairs = this.readKeyValuePairs(stream, ['first', 'second', 'amount']);
+
+        const first: number = parseInt(pairs.get('first')!, 10);
+        assert(first > 0, `handleKerning: first is not greater than 0.`);
+
+        const second: number = parseInt(pairs.get('second')!, 10);
+        assert(second > 0, `handleKerning: second is not greater than 0.`);
+
+        const kerning: number = parseFloat(pairs.get('amount')!);
+
+        fontFace.setKerning(first, second, kerning);
+    }
+
+    protected readKeyValuePairs(stream: Array<string>, mandatoryKeys: Array<string>): StringPairs {
+
+        let key: string;
+        let value: string;
+
+        const pairs: StringPairs = new Map<string, string>();
+
+        for (const s of stream) {
+            const pair = s.split('=');
+            key = pair[0];
+            value = pair[1];
+            pairs.set(key, value);
+        }
+
+        // check if all required keys are provided
+        let valid = true;
+        for (const mandatoryKey of mandatoryKeys) {
+            valid = valid && pairs.has(mandatoryKey);
+        }
+
+        if (!valid) {
+            return new Map<string, string>(); // typescript does not allow writing `new StringPairs()`
+        } else {
+            return pairs;
+        }
+    }
+
+    /**
+     * @param context
+     * @param filename path to a .fnt file
+     * @param headless
+     * @param onImageLoad Callback is called when the glyph atlas is loaded.
+     */
+    load(context: Context, filename: string, headless: boolean, onImageLoad: (() => void)): FontFace {
+        const fontFace = new FontFace(context);
+
+        fetchAsync(filename, (text) => text).then(
```

Comment:
So far so good, but for a fully asynchronouse API design you'd return the promise and not take a callback, i.e. the declaration would look like this: 
```
load(context: Context, filename: string, headless: boolean): Promise<FontFace> {
```
And instead of using `then` you could also use async/await:
```
// add async keyword (!)
async load(context: Context, filename: string, headless: boolean): Promise<FontFace> { 
    ...
    try {
        // promise fulfilled
        let text = await fetchAsync(filename);
        const lines = text.split('\n');
        ...
    } catch (e) {
        // promise rejected
       console.error('ERROR: Could not load font file. filename is: ' + filename);
    }
    ...
}
```
But: there's nothing wrong with passing callbacks and using `then`, it just feels a bit archaic and verbose.

Reply:
Ok, done. I guess you want the same procedure for `handlePage`...?
Why is try-catch better than .then?
And sorry for that archaic coding, first time working with Promises. I suppose I found the old tutorials.

Reply:
Yes, `handlePage` should work similar, since the goal is to get rid of the `onImageLoad` parameter and instead only resolve the promise (return the font face) after everything is loaded.
Does 'page' occur multiple times per file? That would complicate the promise handling a bit, i.e.
```typescript
const promises = [];
for (const l of lines) {
    let line = l.split(' ');
    const identifier = line[0];
    line = line.slice(1);

    if (identifier === 'info') {
        this.handleInfo(line, fontFace);
    } else if (identifier === 'common') {
        this.handleCommon(line, fontFace);
    } else if (identifier === 'page' && !headless) {
        promises.push(this.handlePage(line, fontFace, filename, context));
    } else if (identifier === 'char') {
        this.handleChar(line, fontFace);
    } else if (identifier === 'kerning') {
        this.handleKerning(line, fontFace);
    }
}
await Promise.all(promises);
```
(directly awaiting `this.handlePage` would cause serial loading of the images, but the browser can do several in parallel).

---
try-catch is better because it's the same syntax as for standard synchronous error handling. The whole async/await syntax was introduced so there's not much difference between synchronous and asynchronous code (and to get rid of 'callback hell').

---
No need to be sorry, Javascript moves quite quickly these days :)

Reply:
Well, 'page' *can* occur multiple times per file, but we don't have it tested.

When everything is asynchronous, the rendering begins before the texture is actually loaded. So maybe the wrapping function `loadFont` shouldn't work asynchronously?